File: stripe/src/test/java/com/stripe/android/view/BaseViewTest.java
package com.stripe.android.view;

import android.app.Activity;
import android.content.Intent;
import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.junit.After;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.android.controller.ActivityController;

import java.util.HashMap;
import java.util.Map;

@RunWith(RobolectricTestRunner.class)
public abstract class BaseViewTest<T extends Activity> {
    @NonNull private Class<T> mClazz;
    @NonNull private final Map<Integer, ActivityController<T>> mActivityControllers =
            new HashMap<>();

    protected BaseViewTest(@NonNull Class<T> clazz) {
        this.mClazz = clazz;
    }

    @NonNull
    protected T createActivity() {
        return createActivity(null);
    }

    @NonNull
    protected T createActivity(@Nullable Intent intent) {
        final ActivityController<T> activityController = Robolectric.buildActivity(mClazz, intent)
                .create().start()
                .postCreate(null).resume().visible();
        final T activity = activityController.get();
        mActivityControllers.put(activity.hashCode(), activityController);
        return activityController.get();
    }

    @NonNull
    protected T createStartedActivity() {
        final ActivityController<T> activityController = Robolectric.buildActivity(mClazz)
                .create().start();
        final T activity = activityController.get();
        mActivityControllers.put(activity.hashCode(), activityController);
        return activity;
    }

    /**
     * Call resume() and visible() on an Activity that was created with
     * {@link #createStartedActivity()}.
     */
    protected void resumeStartedActivity(T activity) {
        final ActivityController<T> activityController =
                mActivityControllers.get(activity.hashCode());
        if (activityController != null) {
            activityController.resume().visible();
        }
    }

    @CallSuper
    @After
    public void tearDown() {
        for (ActivityController<T> activityController : mActivityControllers.values()) {
            activityController.pause().stop().destroy();
        }
    }
}

File: stripe/src/test/java/com/stripe/android/view/BaseViewTest.kt
package com.stripe.android.view

import android.app.Activity
import androidx.annotation.CallSuper
import java.util.HashMap
import kotlin.test.AfterTest
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.android.controller.ActivityController

@RunWith(RobolectricTestRunner::class)
internal abstract class BaseViewTest<T : Activity> protected constructor(
    private val activityClass: Class<T>
) {
    private val activityControllers = HashMap<Int, ActivityController<T>>()

    protected fun createStartedActivity(): T {
        val activityController = Robolectric.buildActivity(activityClass)
            .create().start()
        val activity = activityController.get()
        activityControllers[activity.hashCode()] = activityController
        return activity
    }

    /**
     * Call resume() and visible() on an Activity that was created with [createStartedActivity].
     */
    protected fun resumeStartedActivity(activity: T) {
        activityControllers[activity.hashCode()]?.resume()?.visible()
    }

    @CallSuper
    @AfterTest
    open fun tearDown() {
        activityControllers.values.forEach {
            it.pause().stop().destroy()
        }
    }
}
